ADR/ADR036-add-solidqueue-to-forms-runner.md

```diff
@@ -0,0 +1,37 @@
+# ADR036: SolidQueue and ActiveJob for managing background processing
+
+Date: 2024-11-25
+
+## Status
+
+Approved
+
+## Context
+
+Forms will be using AWS SES to deliver emails to form processors with files attached to the emails. For processing of these emails, we looked at Sidekiq and Redis for managing a queue and to give us the ability to retry failed submissions. However with Redis, the details of a submission are ephemeral once a job is taken to be processed. Due to the way Sidekiq uses Redis in the free version is it possible for it to lose jobs if the process crashes. It is very important that we do not lose submissions, so this risk isn't acceptable to us.
+
+We considered just using SQS, but there is a message size limit of 256kb. We cannot be sure that there won’t be form submissions that exceed this size.
```

Comment:
Will we ever have a maximum size, or do we allow any size?

Reply:
There seem to be [workarounds that involve putting the message payload in s3](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-s3-messages.html), rather than the queue itself. The queue contains a reference to the object in s3. It seems to suggest this is up to a limit of 2GB

Reply:
Nice, thanks for this suggestion! I think enforcing a size limit would be difficult - I guess we could add a maximum to free text fields. This probably would have been 8kb per question due to the WAF rules, but now we're letting through larger requests. So theoretically forms could ask for essay-length submissions although I imagine this is very unlikely. So it probably just depends on whether we're happy with that small risk of losing user submissions if they're too large.

I think what is more important, and a reason we probably can't put the submission data just in the SQS message is that the email might bounce - which we hear about later. In this case we need to make sure we're able to retry sending the submission email, so we'd want the data to be maintained until we know it's been successfully delivered.

With the S3 solution you suggested, I think this should be possible. We'd just need a way of linking an email send attempt back to the data in S3 so we can delete the file when it's successful, or retry if it's unsuccessful. Possibly we can use the message ID for this and have it the same as the s3 key - but we'd need to check how this would work.

Reply:
Yeah, and I think Andy's calculations show that we'd still be quite unlikely to hit the SQS limit anyway. Unless there are weird use cases we haven't considered

Reply:
I remember the engagement team mentioning an incredibly long form (100+ page PDF) for setting up a school and I had that in mind as a possible problem form. But that's a very unusual case and I'm not even sure that would hit the limit